v8: update to 6.5.254

Qard opened the pull request:
V8 changed the build tools in 6.x, so it needs to be built with GN now, rather than gyp. I updated everything so it now installs 6.5.254, which is the [latest stable version](https://omahaproxy.appspot.com).

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

DomT4 commented on Formula/v8.rb:
```diff
@@ -3,110 +3,90 @@
-  url "https://github.com/v8/v8-git-mirror/archive/5.1.281.47.tar.gz"
-  sha256 "63c9933227d6912689ea6bc012eea6a1fabaf526ac04bc245d9381e3ea238bf6"
-
-  bottle do
```
Leave the bottle block be. CI will handle it.

Qard replied:
restored

DomT4 commented on Formula/v8.rb:
```diff
@@ -3,110 +3,90 @@
   end
 
   def install
-    # Bully GYP into correctly linking with c++11
```
Has this issue/dance been resolved with newer versions of `v8`?

Qard replied:
V8 doesn't use gyp anymore. They switched to GN, which is why this is not as straightforward as a simple sha update.

DomT4 commented on Formula/v8.rb:
```diff
@@ -3,110 +3,90 @@
+    # see https://www.chromium.org/developers/how-tos/depottools#TOC-Disabling-auto-update
+    ENV["DEPOT_TOOLS_UPDATE"] = "0"
+
+    repo_cache = HOMEBREW_CACHE/"#{name}--v8--gclient/"
```
This feels like a misuse of `HOMEBREW_CACHE`, to be honest. I'll let ILZ chime in though. We might simply be better using `buildpath`.

Qard replied:
I'll wait for more comments on this. I haven't edited a formula before, so I don't really know what the best practices are. 😅

ilovezfs replied:
We do need some kind of caching mechanism here, but probably best not to do the build in the cache. I guess we could copy everything out of the cache into the buildpath after running glient sync.

Qard replied:
I just removed the caching entirely for now, as it didn't actually work properly--reinstalling would fail because the cache was dirty. I'll have to figure out if there's a reasonable way to cache. Unfortunately doing a `gclient sync` to the version you are already synced to will fail, so it's not as straightforward as just copy/pasting the directory. 🤔

DomT4 commented on Formula/v8.rb:
```diff
@@ -3,110 +3,90 @@
+        include.install Dir["include/*"]
+
+        cd output_path do
+          lib.install Dir["lib*.dylib", "icudtl.dat", "d8"]
```
Executables should live in `bin`, not `lib`.

Qard replied:
fixed

DomT4 commented on Formula/v8.rb:
```diff
@@ -3,110 +3,90 @@
 
   test do
-    assert_equal "Hello World!", pipe_output("#{bin}/v8 -e 'print(\"Hello World!\")'").chomp
+    test_basic_script = <<-EOS.strip
```
You're going to need to use `<<~EOS` throughout the test here. Although given how you're using it below you could just inline it, probably.

Qard replied:
I tried just inlining.

DomT4 commented on Formula/v8.rb:
```diff
@@ -3,110 +3,90 @@
+      print(new Intl.DateTimeFormat("en-US").format(date));
+    EOS
+
+    assert_equal "Hello World!", pipe_output("#{lib}/d8 -e '#{test_basic_script}'").chomp
```
`pipe_output` doesn't bother checking the return code by default, so we should handle that here to assure we're getting back `0`.

Qard replied:
I switched to `shell_output`.

Qard commented:
Fixed the ICU failure.

chrmoritz commented on Aliases/v8@5.1:
```diff
@@ -1 +0,0 @@
-../Formula/v8.rb
```
You have to create a new `v8@6.5` alias to make brew audit happy:

```
  * Formula has other versions so create a versioned alias:
      cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases
      ln -s ../Formula/v8.rb v8@6.5
```

Qard replied:
fixed.

chrmoritz commented on Formula/v8.rb:
```diff
@@ -15,98 +17,75 @@ class V8 < Formula
+        :is_debug => false,
+        :is_component_build => true,
+        :v8_use_external_startup_data => false,
+        :v8_enable_i18n_support => true
```
This line should also end with a conna for brew audit:
```C: 69:  9: Put a comma after the last item of a multiline hash.```

Qard replied:
fixed

chrmoritz commented on Formula/v8.rb:
```diff
@@ -15,98 +17,75 @@ class V8 < Formula
+      include.install Dir["include/*"]
+      cd output_path do
+        lib.install Dir["lib*.dylib"]
+        bin.install Dir["d8", "icudt*.dat"]
```
You shouldn't install non executebales to bin. 
```  
* Non-executables were installed to "/usr/local/opt/v8/bin"
    The offending files are:
      /usr/local/opt/v8/bin/icudtl.dat
```
Is it still possible to build v8 against system icu (as is was in good old gyp times).?
If yes, then we should revert the formula back to use system icu and `depends_on "icu4c`. 
If not, than I think the best way forward is to install both `d8` and the icu data files to libexec and only `bin.install_symlink` d8 from there.

Qard replied:
As far as I can tell, `icudtl.dat` is required for i18n features. Good call on the symlink. I'll try that. The `icudtl.dat` file unfortunately has to be in the same path as the `d8` binary, so I just dumped both in `bin`, but the symlink idea is definitely better.

chrmoritz replied:
> As far as I can tell, icudtl.dat is required for i18n features.

Older version of `V8` had the option to build against system icu instead of the one bundled with v8. How to do this with `gyp` is still documented in the now outdated [v8 i18n wiki page](https://github.com/v8/v8/wiki/i18n-support#system-icu).
We used this option in older versions of this formula to build against the [homebrew provided icu4c](https://github.com/Homebrew/homebrew-core/blob/master/Formula/icu4c.rb), but [had to remove it at some point](https://github.com/Homebrew/homebrew-core/pull/16731/commits/52a6d20b9377034b473cf01ed2870237b608489e), because v8 5.1 wasn't (forward) compitable with recent icu releases anymore.
I think building against system icu would be the preffered option here, expecially because it would also solve the `icudtl.dat` needs to be in the same path as `d8` issue. But unfurtunaly I wasn't able to find a `gn` option for it after a quick search and that is why I asked if it is still possible or if they removed support for it with gn.
Otherwise the install to `libexec` and smylink `d8` to `bin` approve should be fine too.

Qard replied:
I updated to do the symlink. I think it's a good balance of still having ICU support even though we can't use the system ICU. I think it's a better option than disabling ICU entirely.

Qard replied:
Hmmm...the symlink approach doesn't seem to work either. I put `d8` and `icudtl.dat` in `libexec`, but when I symlink `d8` into `bin` it seems to make a copy in the cellar `bin` and symlink to that rather than the one in `libexec` so, again, `icudtl.dat` is in the wrong place. 😖

chrmoritz replied:
Ah ... so `d8` doesn't resolve the symlink when looking for `icudtl.dat` files.
Do you know if there is a way to point `d8` to a different path to the icudtl.dat file like node's `NODE_ICU_DATA` environment variable? (We could use a env wrapper script in this case.)

BTW: You need the remove the old `v8@5.1` alias too, so that you have in the end just renamed the alias from `v8@5.1` to `v8@6.5`.

Qard replied:
Yeah, that seems to be the case. It fails if I run `/usr/local/bin/d8` or `/usr/local/Cellar/v8/6.5.254/bin/d8`, but it works if I run `/usr/local/Cellar/v8/6.5.254/libexec/d8` directly.

There's a `--icu-data-file=$somewhere` flag that could be used.

Qard replied:
What's the best way to do the wrapper script idea?

chrmoritz replied:
If there is also a environment variable for it, you could just install to `libexec` and use [`write_env_script`](http://www.rubydoc.info/github/Homebrew/brew/master/Pathname:write_env_script) for it like (assuming the env variable is called `ICU-DATA-FILE`): `bin.write_env_script libexec/"d8", :ICU-DATA-FILE => libexec/"icudtl.dat"`.

Otherwise you have to write a custom wrapper script to bin like:
```ruby
(bin/"d8").atomic_write <<~EOS
  #!/bin/bash
  exec "#{libexec}/d8" --icu-data-file="#{libexec}/icudtl.dat" "$@"
EOS
```

chrmoritz commented:
The test failure with `gjstest` is caused by api incompatibilities between `gjstest` and the v8 6.5 api changes. There is an upstream fix available in https://github.com/google/gjstest/commit/8218472140582e2efd1a9dfb2d7b969610a02d75, but unfortunaly it didn't made it in a stable release yet. Just applying [this patch](https://github.com/google/gjstest/commit/8218472140582e2efd1a9dfb2d7b969610a02d75.patch) to the `gjstest` formula should fix the issue in the meantime.

BTW: Thanks for your effort to bring the v8 formula to a recent version.

Qard commented:
Yeah, I'm hoping that once I get this all cleaned up and ready to land it'll be in a place that I can keep it tracking the latest stable without too much trouble. :)

chrmoritz commented on Formula/v8.rb:
```diff
@@ -15,98 +17,79 @@ class V8 < Formula
+      target_os = [ "mac" ]
+      target_os_only = True
+    EOS
+    system "gclient", "sync", "-vvv", "-j", Hardware::CPU.cores, "-r", v8_version
```
You could add `--no-history` here to only make a shallow clone, because we won`t reusing the repos anyway.

chrmoritz commented on Formula/v8.rb:
```diff
@@ -15,98 +17,79 @@ class V8 < Formula
+      end.join(" ")
+
+      # Build with gn + ninja
+      system "gn gen \"#{output_path}\" --args=\"#{gn_args}\""
```
```ruby
system "gn", "gen", output_path, "--args=\"#{gn_args}\""
```

Qard replied:
Splitting it up results in `ERROR at the command-line "--args":1:1: Expecting assignment or function call.`, which actually happens using `Kernel.system` directly too. I'm not sure if it's a bug in Ruby or in GN, but there's not much I can do about it either way. 🤷‍♂️ 

Is there any way to make an exception for the audit tool here somehow?

chrmoritz commented on Formula/v8.rb:
```diff
@@ -15,98 +17,79 @@ class V8 < Formula
+        # Install d8 and icudtl.dat in libexec and symlink
+        # because they need to be in the same directory.
+        libexec.install Dir["d8", "icudt*.dat"]
+        bin.install_symlink libexec/"d8"
```
I can confirm, that replacing this line with a wrapper script setting `--icu-data-file` will work:
```ruby
mkdir_p bin
(bin/"d8").atomic_write <<~EOS
  #!/bin/bash
  exec "#{libexec}/d8" --icu-data-file="#{libexec}/icudtl.dat" "$@"
EOS
```

chrmoritz commented on Formula/v8.rb:
```diff
@@ -3,8 +3,10 @@
+  version "6.5.254"
+
+  # NOTE: Hack because we need depot_tools and don't want to pull v8 twice.
+  url "https://chromium.googlesource.com/chromium/tools/depot_tools.git"
```
I think it can actually work. What do you think about this (mis)use of the gclient cache?

https://github.com/chrmoritz/homebrew-core/commit/d82b9e0c30b781a2d87ec925ddfb37817e9d05c8

Qard replied:
I think it could work. I'll give it a try...

fxcoudert commented:
```
==> Checking out tag 6.5.254
warning: refname '6.5.254' is ambiguous.
Error: 6.5.254 tag should be 73c55f57fe8506011ff854b15026ca765b669700
but is actually 62df57d00d03a1a76e836178b1a4ac3ab1ab79ab
```

Qard commented on Formula/v8.rb:
```diff
@@ -3,8 +3,10 @@
+  head "https://chromium.googlesource.com/v8/v8.git"
+  url "https://chromium.googlesource.com/v8/v8.git",
+    :revision => "73c55f57fe8506011ff854b15026ca765b669700",
+    :tag => "6.5.254"
```
fixed

Qard commented on Formula/v8.rb:
```diff
@@ -15,98 +17,237 @@ class V8 < Formula
-      rm ["libgmock.a", "libgtest.a"]
-      lib.install Dir["lib*"]
-      bin.install "d8", "mksnapshot", "process", "shell" => "v8"
+    ENV.prepend_path "PATH", buildpath
```
fixed

Qard commented on Formula/v8.rb:
```diff
@@ -15,98 +17,237 @@ class V8 < Formula
+    (buildpath/"#{base}-llvm--project-libcxxabi").install resource("libcxxabi")
+    (buildpath/"#{base}-llvm--project-cfe-tools-clang--format").install resource("clang-format")
+    (buildpath/"#{base}-v8-deps-third_party-mozilla--tests").install resource("mozilla-tests")
+    (buildpath/"#{base}-v8-deps-third_party-benchmarks").install resource("benchmarks")
```
fixed

Qard commented:
Hey, sorry for the silence...I'll get back to this soon. 😅

Qard commented:
I've added another commit bumping v8 to 6.6 now, since that's the latest stable version now. I can squash it into the previous v8 commit, if you prefer.

Qard commented:
What's the difference between aliases and formulas that are versioned in their name? In my experimenting, it seems like dependencies will automatically use the latest of that exact named formula. (which I guess is why I see lots of `python@2` stuff?) Should I actually be doing the same thing here and just making a `v8@6.6` formula, so it doesn't break stuff that depends on `v8`? V8 changes rapidly, so pretty much any update is likely to be a breaking change at the API or ABI level.

mjerris commented:
would it be possible to add a pc file to this formula.  thinking something along the lines of:

add this to install:
(lib+"pkgconfig/v8.pc").write pc_file  

new function:
  def pc_file; <<~EOS                                                                                                                                           
    prefix=#{opt_prefix}                                                                                                                                        
    exec_prefix=${prefix}                                                                                                                                       
    libdir=${exec_prefix}/lib                                                                                                                                   
    includedir=${prefix}/include                                                                                                                                
                                                                                                                                                                
    Name: v8                                                                                                                                                
    Description: V8 JavaScript Engine                                                                                                                       
    Version: #{version}                                                                                                                                         
    Libs: -L${libdir} -lv8 -lv8_libbase -lv8_libplatform                                                                                                                             
    Cflags: -I${includedir}                                                                                                                                     
    EOS                                                                                                                                                         
  end

mjerris commented:
Also, any update on where this stands and what else needs to be figured out still?

Qard commented:
Yeah, I can see about adding a pc file when I have some time. Is that a thing that other formulas do?

As for status, I'm a bit uncertain if this should use aliased versions or the named version thing like `python@2`. I'm also unsure about how to deal with `gn gen` not working when its split up. There also seemed to be another issue still with gjstest that the patch didn't fix, so it's a bit blocked on that too. If I made this update as a named version, like `v8@6.6`, then the gjstest issues shouldn't matter.

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
+  # depot_tools/GN require Python 2.7+
+  depends_on "python@2" => :build
+
+  depends_on "readline" => :optional
```
The formula has 1,073 installs per month, exactly zero of which use the `--with-readline` option. Let's remove it (so no users ever have to build this from source!).

Qard replied:
fixed

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
+      :revision => "f6c578a10ea707b1a8ab0b88943fe5115ce2b9be"
+  end
+
+  # not building on Yosemite
```
URL is enough comment here, remove `not building on Yosemite`

Qard replied:
fixed

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
-      rm ["libgmock.a", "libgtest.a"]
-      lib.install Dir["lib*"]
-      bin.install "d8", "mksnapshot", "process", "shell" => "v8"
+    v8_version = head? ? "HEAD" : version
```
`v8_version` is not necessary once head is removed.

Qard replied:
fixed

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
+    v8_version = head? ? "HEAD" : version
+    arch = MacOS.prefer_64_bit? ? "x64" : "x86"
+    output_name = "#{arch}.release"
+    output_path = "out.gn/#{output_name}"
```
`output_path  = "out.gn/x64.release"`
All archs are 64-bit now.

Qard replied:
fixed

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
+    base = "gclient-cache/chromium.googlesource.com"
+    (buildpath/"#{base}-v8-v8").install Dir["*"]
+
+    depot_tools = resource("depot_tools")
```
Extra variable not necessary, makes reading harder (too many depot_tools).
Just inline `resource("depot_tools")` in the two places where it's used.

Qard replied:
fixed

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
+    (buildpath/"depot_tools").install depot_tools
+    ENV.prepend_path "PATH", buildpath/"depot_tools"
+
+    # This env variable used by gclient to prevent depot_tools to update depot_tools on every call
```
`s/This env variable used by gclient to//`

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
+      target_os_only = True
+    EOS
+
+    system "gclient", "sync",
```
Indent with more whitespace

Qard replied:
How much? Not sure what it's _supposed_ to look like. Do you align subsequent lines with the first argument or something like that?

fxcoudert replied:
All lines indented to the _second_ argument (see other formulas for examples)

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
+
+    system "gclient", "sync",
+      "--cache-dir=#{buildpath}/gclient-cache",
+      "-j", Hardware::CPU.cores,
```
Use `ENV.make_jobs` instead

Qard replied:
fixed

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +17,217 @@ class V8 < Formula
+      # Build with gn + ninja
+      system "gn gen \"#{output_path}\" --args=\"#{gn_args_string}\""
+      system "ninja",
+        "-j", Hardware::CPU.cores,
```
`ENV.make_jobs`

Qard replied:
fixed

fxcoudert commented:
@Qard if upstream doesn't ship with a `.pc` file, we typically don't add it, unless there is strong evidence that other vendors (Debian, for example) add their own and external software uses it.

fxcoudert commented:
Once changes are made, please squash every change to the `v8.rb` file into a single commit, please.

chrmoritz commented:
Also we have to decide how we move forwards with `gjstest`. It doesn't build against more recent versions of v8 anymore (the upstream patch we tried to apply unfortunately didn't fix all bulld errors), the last release is from 2015 and it's not in the 1000 top most installed formulas. Either let it depend on an older versioned `v8` formula or remove it completely?

fxcoudert commented:
Regarding `gjstest`, there are apparently fixing it: https://github.com/google/gjstest/commit/8218472140582e2efd1a9dfb2d7b969610a02d75

If they don't provide a new release, maybe a patch is the way to go. But this shouldn't hold v8 update, really.

chrmoritz commented:
@fxcoudert: We've already tried to apply this patch in https://github.com/google/gjstest/commit/8218472140582e2efd1a9dfb2d7b969610a02d75. Unfortunatly it doesn't fix all of the build errors with v8 6.5+ as likely a few more deprecated API's got removed from v8 6.5 after this patch was created.

mjerris commented:
@chrmoritz do you have the build output with that patch.  Just went through the exercise of updating our code to support 6.6 yesterday so what needs to be done is fresh.  Might be able to assist with this.

chrmoritz commented:
@mjerris With the formulae version from this PR atm: https://gist.github.com/chrmoritz/584e5683b9d4f9c91b8414848b187acd#file-01-make-L135

chrmoritz commented:
Same error with only `6.6.346.26` which is indeed strange.

Qard commented:
The undefined symbol for `typeinfo for v8::ArrayBuffer::Allocator` is because the shape of the Allocator class changes in 6.6 and gjstest implements the old interface. I got it working by just using the default allocator in a local checkout, instead of their custom allocator, as they actually do the same thing. I was able to fix the `v8::platform::CreateDefaultPlatform(int)` error by using `NewDefaultPlatform` instead. I'm not sure why yet, but the linkage for `CreateDefaultPlatform` didn't seem to work properly in 6.6. 🤔

Qard commented:
I should point out that V8 historically has made breaking changes regularly in point releases, so it's probably best the formula be designed in such a way that one has to depend on more specific versions. Between 6.5 and 6.6 there were a bunch of breaking changes, so it's likely upgrading just between those two versions would break a lot of apps. 😅

chrmoritz commented:
> I'm not sure why yet, but the linkage for CreateDefaultPlatform didn't seem to work properly in 6.6.

If I remember correctly, this was already broken in 6.5 too (which caused the build to fail there).

Qard commented on Formula/v8.rb:
```diff
@@ -3,8 +3,10 @@
   homepage "https://github.com/v8/v8/wiki"
-  url "https://github.com/v8/v8-git-mirror/archive/5.1.281.47.tar.gz"
-  sha256 "63c9933227d6912689ea6bc012eea6a1fabaf526ac04bc245d9381e3ea238bf6"
+  head "https://chromium.googlesource.com/v8/v8.git"
```
fixed

Qard commented:
I added another commit with a kind of gross hack around the audit failure. Unfortunately it's not possible to use the multi-input form of the `system` function that the audit requires as Ruby doesn't seem to handle the args string for gn correctly. 🤔

Alhadis commented:
> Between 6.5 and 6.6 there were a bunch of breaking changes, so it's likely upgrading just between those two versions would break a lot of apps

Versioning done right.

<img src="https://user-images.githubusercontent.com/2346707/39685488-397bb9cc-5206-11e8-96ab-63c4d25d6a8a.png" alt="Now THAT'S what I call Semantic Versioning" width="500" align="center" />

Qard commented:
V8 doesn’t use semver. https://github.com/v8/v8/wiki/Version-numbers

chrmoritz commented:
> as Ruby doesn't seem to handle the args string for gn correctly.

I doubt that it is the arg string who is causing the problem, because `system "gn", "gen \"#{output_path}\" --args=\"#{gn_args_string}\""` does already break it for me.

mjerris commented:
@Qard i thought you said you had a patch that had gjstest working with 6.6?

Qard commented:
Not so much a patch yet, no. I did a manual check out of the project and made a bunch of local edits. It throws away their entire allocator though, so it'd take a bunch of changes. I also didn't really test it at all other than just verifying it was able to build. Should probably be made into a proper PR.

In any case, do we still *want* to do anything about gjstest here? Seems like it's not used much anyway, nor does it seem very actively maintained...

fxcoudert commented on Formula/v8.rb:
```diff
@@ -15,98 +16,211 @@ class V8 < Formula
+
+      # Build with gn + ninja
+      gen_command = "gn gen \"#{output_path}\" --args=\"#{gn_args_string}\""
+      system gen_command
```
`system "gn", "gen", "#{output_path}", "--args=#{gn_args_string}"`
will be clearer, without quoting quotation marks

Qard replied:
The quotes are required. The args string will have spaces.

fxcoudert replied:
Calling `system` with multiple args, instead of a single string, will enforce the quoting automatically.

Qard replied:
I tried it like that, but it didn't work.

Qard commented:
I'm traveling for the next two weeks. I'll have to put this on hold until I get back. 😅

Qard commented:
Finally got back to this...

I removed the unused `target_os_only` flag, fixed the resource revisions, and changed the `gn gen` call to use the split up form of `system` call. Is there anything else that still needs to be done?

Qard commented:
It appears to be passing all the tests now, other than gjstest being broken. What should I do about gjstest?

DomT4 commented:
Are we behind upstream again? I know Chrome is shipping `6.7.288.44` currently.

Qard commented:
Yep. Stable is currently at 6.7.288.46. I can update later, if we want to keep trying to track while the overall PR is still under review. Might be better to just get what we have now landed though, and then worry about updating later. 🤔

DomT4 commented:
It's difficult walking the line on updating `v8`. It's been a complete nightmare to get anywhere near done, yet even before they changed the entire build system that remained true, so the temptation is to try and ship the latest possible if it's going to be a pain either way. 

> What should I do about gjstest?

At the risk of making myself unpopular upstream, it could be worth just removing it for now. It can always make a comeback when/if it becomes compatible again, but right now it's _way_ less popular than `v8` and I'm personally not super 👍on vendoring an entire old `v8` for anything.

DomT4 commented on Formula/v8.rb:
```diff
@@ -15,98 +16,213 @@ class V8 < Formula
-    url "https://chromium.googlesource.com/external/gyp.git",
-        :revision => "4ec6c4e3a94bd04a6da2858163d40b2429b8aad1"
+  resource "depot_tools" do
+    url "https://chromium.googlesource.com/chromium/tools/depot_tools.git"
```
This should probably be pinned to a `revision`.

Qard replied:
Google recommends always using the latest version of depot_tools. In fact, I believe it will actually update itself. I can pin it if we really want to do that, but I'm not sure it actually makes sense in this case.

DomT4 replied:
It mostly just makes me a little nervous that we're intentionally introducing a permanent inconsistency into the build. `depot_tools` is updated regularly enough that someone building from source today could end up with a slightly different set of build bugs or behaviours than someone building from source in 3 days time.

I guess if it's going to update itself anyway & we can't stop that then 🤷‍♂️ , but yeah, deliberate inconsistency in the build process sets me on edge a tiny bit.

Qard replied:
Yeah...that's just how Google does it though...I'm doing my best to massage their build process into something usable in homebrew. They are a bit at odds because the development pace of V8 is much, much faster. 😅

DomT4 replied:
Aye. As (IIRC) the last person to kick the `v8` formula through a significant rewrite I can empathise with the pain of handling anything `v8` in this context 🙃. If we're stuck with it we're stuck with it; unless it causes us hassle I guess I could learn to live with it here.

ilovezfs replied:
@DomT4 is correct that this needs to be pinned to a specific version. We do not want to make up one morning to a broken build in the stable spec just because of some random change in depot tools.

Qard replied:
I can pin it, if you insist. It won’t actually do anything though, since depot tools will update itself automatically. 🤷‍♂️

ilovezfs replied:
That makes me wonder whether we should even be packaging v8.

chrmoritz replied:
You can disable the auto updating by setting the `DEPOT_TOOLS_UPDATE=0` environment variable.
Updating `depot_tools` to the latest version every time we update v8 should hopefully be enough the ensure compatibility.

DomT4 commented on Formula/v8.rb:
```diff
@@ -15,98 +16,213 @@ class V8 < Formula
-  resource "icu" do
-    url "https://chromium.googlesource.com/chromium/deps/icu.git",
-        :revision => "c291cde264469b20ca969ce8832088acb21e0c48"
+  resource "chromium-buildtools" do
```
Are these listed somewhere upstream as to which commit these resources should be pegged to or are we just using the tip of each branch at the time of update?

Qard replied:
The `DEPS` files in the various repos may or may not include dependencies on other things. Currently, V8 itself has [these deps](https://github.com/v8/v8/blob/6.6.346.26/DEPS#L13-L66), and the buildtools dependency of V8 has [these deps](https://chromium.googlesource.com/chromium/buildtools.git/+/2888931260f2a32bc583f005bd807a561b2fa6af/DEPS).

DomT4 replied:
Alright. Let's retain a note somewhere near the top of the resources of where we're pulling the commits from, so it's easier for others to update in future.

DomT4 commented on Formula/v8.rb:
```diff
@@ -15,98 +16,213 @@ class V8 < Formula
+        # because they need to be in the same directory.
+        libexec.install Dir["d8", "icudt*.dat"]
+        mkdir_p bin
+        (bin/"d8").atomic_write <<~EOS
```
Why `atomic_write` instead of `.write`? There's no chance of `bin/d8` already existing, no?

Qard replied:
Not sure exactly. That was copy/pasted from a comment from @chrmoritz earlier in the PR. https://github.com/Homebrew/homebrew-core/pull/25786#discussion_r179010862

DomT4 replied:
Let's try a `.write`, not least because then the `mkdir_p bin` can go 👋.

ilovezfs commented:
looks like pr is conflicted

ilovezfs commented:
Can't depot tools fetch these resources at the correct revisions itself? I'm really not comfortable with a formula that has a pile of manually written resource blocks pinned to various revisions copied out of a file that should be parsed by a tool.

Qard commented:
Yes, depot tools will fetch all of these itself. However, without making resources for them, they will be refetched with each install, even if they are the same commit and therefore unchanged, which makes the install _really_ long. An install without anything cached takes about 2.5 hours for me. 🙄

ilovezfs commented:
Depot tools should be able to be pointed at the `HOMEBREW_CACHE` for this without using resource blocks. See https://github.com/Homebrew/homebrew-core/blob/master/Formula/sync_gateway.rb#L26-L28

chrmoritz commented on Formula/v8.rb:
```diff
@@ -15,98 +16,213 @@ class V8 < Formula
+        },
+      ]
+      target_os = [ "mac" ]
+      target_os_only = True
```
adding a `cache_dir = "#{HOMEBREW_CACHE}/gclient_cache"` here should tell `gclient` to cache it's git checkouts inside `HOMEBREW_CACHE`, see https://dev.chromium.org/developers/how-tos/depottools#TOC-.gclient-file

DomT4 replied:
We'll need a matching `brew` PR to tweak `cleanup` to handle that.

brandonros commented:
Isn't 6.7.290 the latest stable version?

DomT4 commented:
Hush bot.

chrmoritz commented:
@Qard: Looks like you haven't much time recently for this PR. Would you mind if I try to continue fixing the remaining issues to get this finally merged?

Unfortunately is seems that `gclient` has changed it cache layout and our hack to prefill it with homebrew downloaded stuff doesn't work anymore. Looks like we now have basically 2 options (both are tested and working builds).

1. https://github.com/chrmoritz/homebrew-core/blob/b998d4f6c9e03def80790751bd83a6b50777c451/Formula/v8.rb Use `gclient` to download everything (similar to as @ilovezfs [suggested](https://github.com/Homebrew/homebrew-core/pull/25786#issuecomment-399764376)), but even the `v8` source code itself. The main url of the formula would be a latest revision of `depot_tools` git with a fake v8 version and we tell `gclient` to download v8 for this specific version/tag and all its dependencies on it's own. With this we basically given the source code management completely out of our hands and depend 100% on `gclient` there.
2. https://github.com/chrmoritz/homebrew-core/blob/325e59d2581a07c896eaef51bb019ff1420e34ed/Formula/v8.rb Use `gclient` not at all. Instead parse the `DEPS` file with a script and reimplement everything `gclient` would do [with formula logic](https://github.com/chrmoritz/homebrew-core/blob/325e59d2581a07c896eaef51bb019ff1420e34ed/Formula/v8.rb#L135-L172). I've quickly hacked a [python2 script](https://gist.github.com/chrmoritz/ba453f27676da13928e237010491b0af) for this together, which has generated this formula (needs `pip2 install jinja2`). It still could need some work, which I would invest, if we decide that this is the way to go.

Qard commented:
Yeah, sorry...I just haven't had the time or motivation for this recently. I'll happily contribute to refreshing the minor updates in the future, but getting the gn rewrite through has been a bit of a pain. 🙄

DomT4 commented:
> Looks like we now have basically 2 options (both are tested and working builds for latest v8 6.8).

IMO option 1 is the least grim, but upstream are definitely starting to push the boundaries of what is considered generally acceptable behaviour for a formula.

chrmoritz commented:
> Parsing the DEPS file relies too much on implementation details that could change at any time, and the V8 team doesn't have a tendency to think about stability of internals or repo layout...

I don't see it that pessimistic. If you parse the `DEPS` file as a python file (and with this I mean actually running the python interpreter on it, parsing it with regexes would be insane) and implementing their magic Var function there isn't much stuff left which could break the parser script. The only concern I have is, that the could introduce new conditions at any time and that is why I'm already blacklisting them instead of white listing, so that in the worst case we would only end up downloading to much unnecessary stuff. Also if they change anything that would require adjusting the script from option 2 it's quite likely that the same change would also require adjusting the formula for option 1.

> but upstream are definitely starting to push the boundaries of what is considered generally acceptable behaviour for a formula

Yeah, that why I've prototyped option 2. The funny thing is, that our copy of `depot_tools` is only used to download the v8 sources and the sources of it's dependencies according to the `DEPS` files, including a (likely older) copy of `depot_tools` itself (as specified in v8 `DEPS` file), which is than used during the build instead of our copy of `depot_tools`. So the question is, if option 1 is still acceptable, if not we have already an alternative in preparation.

DomT4 commented:
I've asked other maintainers to weigh in here. My preference would be option one but it's a very tentative preference.

fxcoudert commented:
My opinions would be: option 1 > move to cask > option 2.

chrmoritz commented:
I was able to fix `gjstest` compatibility with modern `v8` versions by applying two additional upstream patches (upstream development is pretty much HEAD-only recently).

My https://github.com/Homebrew/homebrew-core/compare/master...chrmoritz:v8-gclient branch should be pretty much ready to go now if we want to go with option 1. @Qard feel free to update this PR with the contents of this branch or should I create a new own one?

MikeMcQuaid commented:
My $0.02: how much is our use of `resource` blocks here creating rather than solving problems? If we just let V8 sort out it's own dependencies and submodules does that make it more reliable.

chrmoritz commented:
> My $0.02: how much is our use of resource blocks here creating rather than solving problems? If we just let V8 sort out it's own dependencies and submodules does that make it more reliable.

The only concern I have personally with option 1 is, that homebrew isn't used for source code controll at all in this case. Homebrew downloads exactly 0% of the code, which is used during the build and the only thing which specify what souce code we actuall get is the fake version which is passed in as a parameter to `gclient`, which is part of `depot_tools`, which we check out as the main url in the formula in always the latest revision (as it is an unversioned tool).
Then `gclient` will check out the `DEPS` for for this specifc v8 tag (we pass in through the fake version) and all recursice `DEPS` files and will download 100% of the source code on it's own, bypassing homebrews integrety checks. It will also download a second (likely older copy) of `depot_tools` itself (as specified in the `DEPS` file), which is then used during the build instead of our own dowloaded `depot_tool` copy.

But if you are all fine with that I'm more the happy to go with option 1 to avoid all the hassle with the `gclient`-less resource handling. I just want to be sure, that you are aware of the caveats and that there is in theory an alternative for that.

DomT4 commented:
> will download 100% of the source code on it's own, bypassing homebrews integrety checks.

That applies to pretty much everything that uses `maven`, `npm`, `cargo`, `cabal`, etc as well already. We do our best to enforce our own integrity checks as broadly as possible but at some point we're forced to surrender some trust/control when using other package managers inside of this package manager. I think `gclient` falls under the same sort of loose banner.

I dislike both options, honestly, but option 1 at least feels like if we all got hit by a bus tomorrow whilst carrying our Macs someone else would know how to maintain the formula without resorting to reanimating the dead.

chrmoritz commented:
> That applies to pretty much everything that uses `maven`, `npm`, `cargo`, `cabal`, etc as well already.

Yeah, but the difference is that in the cases you've mentioned we at least downlaod the main source file of the library/tool and let `maven`, `npm`, `cargo`, `cabal`, etc just download all the dependencies for us. Here we would let `depot_tools` download the main (`v8`) source too.
But yes, I see your point in having to maintain the formula generating script somehow, so option 1 seems best to me with regard to this.

MikeMcQuaid commented:
> Yeah, but the difference is that in the cases you've mentioned we at least downlaod the main source file of the library/tool and let maven, npm, cargo, cabal, etc just download all the dependencies for us. Here we would let depot_tools download the main (v8) source too.
> But yes, I see your point in having to maintain the formula generating script somehow, so option 1 seems best to me with regard to this.

How about as a middle-ground compromise we use a resource block and/or url for the main v8 source but not for anything else?

Alhadis commented:
Is forking `depot_tools` not an option? If memory serves, the entire suite is just a Git repository of various tools... so unless it's riddled with blobs of precompiled, closed-source code, it should (theoretically) be possible to maintain a fork. Wouldn't it?

This is less a suggestion than a question, because I'd like to port Electron to OpenBSD someday, which undoubtedly involves similar headaches with `depot_tools`... 😓

MikeMcQuaid commented:
Long-term forks maintained by Homebrew are not an option for us (or really a good option for any non-large-corporation), sorry.

chrmoritz commented:
> ow about as a middle-ground compromise we use a resource block and/or url for the main v8 source but not for anything else?

@MikeMcQuaid: It's not possible anymore, because our `gclient_cache` hack doesn't work anymore cause of recent changes to the cache layout. Or lets say at least converting a simple git checkout to their new cache format would be way more complex than going with option 2. We could in theory always use a older copy of `depot_tools`, but this will likely break at some day and tbh. the hack was just for giving `gclient` a precloned git repo, on which it would still run a full `git fetch` and `git checkout <commit>` (as in `DEPS` file)  every time on it (assuming git commit was right, this would just be a noop outside the `.git` folder).

> Is forking depot_tools not an option?
@Alhadis: But the maintaining overhead would be much bigger than the maintaining overhead for the `gclient`-less (option 2) generating script, which was the main argument against it.

chrmoritz commented:
I've opened #30961 with the proposed option 1 , as @Qard said he's fine with me taking this PR over in https://github.com/Homebrew/homebrew-core/pull/25786#issuecomment-411203520.

DomT4 commented:
Closing in favour of https://github.com/Homebrew/homebrew-core/pull/30961. Thank you @Qard for being so willing to work on this for so long; I fully appreciate trying to update `v8` outside of its usual update mechanism is not... fun.

Alhadis commented:
Amen to what @DomT4 said, man. 😞  Google's build system took its toll upon me, too. I no longer wonder why they removed the "Don't be evil" from their company's legal clauses ...

Qard commented:
I'm glad someone kept this going. Sadly, I haven't had the time for it recently. 😓

chrmoritz commented:
> I'm glad someone kept this going. Sadly, I haven't had the time for it recently. 😓

No problem, we all have a life outside of open source contributing. Thanks for all your work so far on this PR.
